dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriterImpl.kt

```diff
@@ -0,0 +1,260 @@
+import org.jetbrains.kotlinx.dataframe.api.convertToFloat
+import org.jetbrains.kotlinx.dataframe.api.convertToInt
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalDate
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalDateTime
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalTime
+import org.jetbrains.kotlinx.dataframe.api.convertToLong
+import org.jetbrains.kotlinx.dataframe.api.convertToShort
+import org.jetbrains.kotlinx.dataframe.api.convertToString
+import org.jetbrains.kotlinx.dataframe.api.forEachIndexed
+import org.jetbrains.kotlinx.dataframe.api.map
+import org.jetbrains.kotlinx.dataframe.exceptions.CellConversionException
+import org.jetbrains.kotlinx.dataframe.exceptions.TypeConverterNotFoundException
+
+/**
+ * Save [dataFrame] content in Apache Arrow format (can be written to File, ByteArray, OutputStream or raw Channel) with [targetSchema].
+ * If [dataFrame] content does not match with [targetSchema], behaviour is specified by [mode], mismatches would be sent to [mismatchSubscriber]
+ */
+internal class ArrowWriterImpl(
+    override val dataFrame: DataFrame<*>,
+    override val targetSchema: Schema,
+    override val mode: ArrowWriter.Companion.Mode,
+    override val mismatchSubscriber: (ConvertingMismatch) -> Unit = ignoreMismatchMessage
+) : ArrowWriter {
+
+    private val allocator = RootAllocator()
+
+    private fun allocateVector(vector: FieldVector, size: Int) {
+        when (vector) {
+            is FixedWidthVector -> vector.allocateNew(size)
+            is VariableWidthVector -> vector.allocateNew(size)
+            else -> TODO("Not implemented for ${vector.javaClass.canonicalName}")
+        }
+    }
+
+    private fun infillWithNulls(vector: FieldVector, size: Int) {
+        when (vector) {
+            is BaseFixedWidthVector -> for (i in 0 until size) { vector.setNull(i) }
+            is BaseVariableWidthVector -> for (i in 0 until size) { vector.setNull(i) }
+            else -> TODO("Not implemented for ${vector.javaClass.canonicalName}")
+        }
+        vector.valueCount = size
+    }
+
+    private fun convertColumnToTarget(column: AnyCol?, targetFieldType: ArrowType): AnyCol? {
+        if (column == null) return null
+        return when (targetFieldType) {
+            ArrowType.Utf8() -> column.map { it?.toString() }
+            ArrowType.LargeUtf8() -> column.map { it?.toString() }
+            ArrowType.Binary(), ArrowType.LargeBinary() -> TODO("Saving var binary is currently not implemented")
+            ArrowType.Bool() -> column.convertToBoolean()
+            ArrowType.Int(8, true) -> column.convertToByte()
+            ArrowType.Int(16, true) -> column.convertToShort()
+            ArrowType.Int(32, true) -> column.convertToInt()
+            ArrowType.Int(64, true) -> column.convertToLong()
+//            ArrowType.Int(8, false), ArrowType.Int(16, false), ArrowType.Int(32, false), ArrowType.Int(64, false) -> todo
```

Comment:
what to do with these todos?

Reply:
Hmm, still not very comfortable having a lot of extra TODOs in the main branch. @koperagen, what do you think?

Reply:
Maybe it could be written like
```kotlin
when (vector) {
    is VarBinaryVector,
    is LargeVarBinaryVector,
    ...,
    is StructVector -> throw NotImplementedError(...)
    else -> throw IllegalArgumentException(...)
}
```
That way there's no commented-out code with lots of TODO's (although, 1 TODO in a comment would be fine to indicate there's work to be done here :), but they would definitely have to be gone for a 1.0 release one day). Plus a user would know if the type could be supported in the future or if something else is going wrong.
What do you think?

Reply:
Commented code is removed.

Splitting NotImplementedError and IllegalArgumentException might be not correct if some other types would be added in future. I would like to implement ListVector and FixedSizeListVector (aka jagged martix, flat matrix) someday but not in the current customer's project, sorry

Reply:
I see, that's right indeed